### PR TITLE
Object data storage

### PR DESCRIPTION
**Pull request by Hjortsberg**
Need to find a smart way to make the governer that republishes and evicts values.

### CODE REVIEW

**Comment by Hjortsberg** [comment]
Current performance of ```AddItem()``` is 352 ns/op single thread (if i understand the Benchmarks correct). 
Which amounts to the possibility of inserting 2,84 million data entry's / second.

**Comment by Hjortsberg** [comment]
@optmzr , @edvinnn, I think the first iteration might be ready to put to test.
Comments plz. :octocat:

File: store/store.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
+package store
+
+//import "log"
```
Comment (optmzr) [inline store/store.go]: Remove

File: store/store.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import "time"
+import "golang.org/x/crypto/blake2b"
+import "sync"
+import "fmt"
+
+type NodeID [32]byte
```
Comment (optmzr) [inline store/store.go]: `node.ID` has been merged, so you can change all occurrences of `NodeID` to `node.ID`

File: store/store.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+	republish time.Time
+	origPub   NodeID
+}
+
+type localItem struct {
+	value     string
+	repubTime time.Time
+}
+
+type items struct {
```
Comment (optmzr) [inline store/store.go]: Maybe name this one to `remoteItems` to match `localItems`?

File: store/store.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+
+type replicate struct {
+	sync.RWMutex
+	time time.Time
+}
+
+type Database struct {
+	items      items
+	localItems localItems
+	ch         chan (localItem)
```
Comment (optmzr) [inline store/store.go]: Remove parantheses

File: store/store.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+
+func (db *Database) getReplicate() time.Time {
+	db.replicate.RLock()
+	time := db.replicate.time
+	db.replicate.RUnlock()
+
+	return time
+}
+
+// TODO: Max 1000 chars. Truncate input string.
```
Comment (optmzr) [inline store/store.go]: Add truncation before merging?

```go
func truncate(s string) string {
    if len(s) > 1000 {
        return s[:1000]
    }
    return s
}
```

File: store/store.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+func (db *Database) AddItem(value string, origPub NodeID) error {
+	t := time.Now()
+	expire := t.Add(db.tExpire)
+	republish := t.Add(db.tRepublish)
+
+	key := Key(blake2b.Sum256([]byte(value)))
+
+	newItem := item{}
+
+	newItem.value = value
```
Comment (optmzr) [inline store/store.go]: Use `truncate` function, like:
```go
newItem.value = truncate(value)
```

File: store/store.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+	newItem.expire = expire
+	newItem.republish = republish
+	newItem.origPub = origPub
+
+	// TODO: Handle duplicate keys.
```
Comment (optmzr) [inline store/store.go]: We decided to not handle this after the talk with Erik, right?

File: store/store.go [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+	if !found {
+		err = fmt.Errorf("store: GetItem, no item matching key: %x", key)
+		return
+	}
+
+	return requestedItem, nil
+}
+
+/*
+func (db *Database) GetRepubTime(key Key) (repubTime time.Time, err error) {
```
Comment (optmzr) [inline store/store.go]: I suppose you should remove this one?

File: store/store.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	return localItem, nil
+}
+
+func (db *Database) evictItem(key Key) {
+	db.items.Lock()
+	delete(db.items.m, key)
+	db.items.Unlock()
+}
+
+// Start this as a Goroutine at node start.
```
Comment (optmzr) [inline store/store.go]: Comments in Go follows a convention where the first word must be the function name, like:
```
// itemHandler checks for expired items every second and remove them if they're outdated.
// This function should be run as a goroutine.
```

File: store/store.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+
+		for _, key := range evictees {
+			db.evictItem(key)
+		}
+	}
+}
+
+// Start as Goroutine.
+// Function is responsible for republishing data that should persist on storage network.
+func (db *Database) republishHandler() {
```
Comment (optmzr) [inline store/store.go]: Fix the comment here as with `itemHandler`

File: store/store.go [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+	for {
+		timer := time.NewTimer(time.Second * 1)
+		<-timer.C
+
+		now := time.Now()
+		replicate := now.After(db.getReplicate()) // This line is a suspect if something doesnt work.
```
Comment (optmzr) [inline store/store.go]: Why is it suspect?

File: store/store_test.go [hunk 12]
```diff
@@ -0,0 +1,263 @@
+package store
+
+import "testing"
+import "time"
+
+//import "strconv"
+//import "fmt"
+//import "golang.org/x/crypto/blake2b"
+//import "encoding/hex"
```
Comment (optmzr) [inline store/store_test.go]: Remove these commented lines.

File: store/store_test.go [hunk 13]
```diff
@@ -0,0 +1,263 @@
...
+	db, _ := NewDatabase(time.Second*86400, time.Second*3600, time.Second*86400)
+
+	testVal := "q"
+
+	var testNodeID NodeID
+	copy(testNodeID[:], "w")
+
+	err := db.AddItem(testVal, testNodeID)
+	if err != nil {
+		t.Errorf("some error in AddItem that is not yet defined")
```
Comment (optmzr) [inline store/store_test.go]: So `db.AddItem` cannot error? Just remove the return type instead and skip this check here

File: store/store_test.go [hunk 14]
```diff
@@ -0,0 +1,263 @@
...
+		t.Errorf("NodeID stored does not match original NodeID.\nExpected: %x\nGot: %x", testNodeID, storedTestItem.origPub)
+	}
+
+	if storedTestItem.expire.IsZero() || storedTestItem.republish.IsZero() {
+		t.Errorf("Expire or republish time is not set")
+	}
+}
+
+func BenchmarkAddItem(b *testing.B) {
+
```
Comment (optmzr) [inline store/store_test.go]: Remove newline

File: store/store_test.go [hunk 15]
```diff
@@ -0,0 +1,263 @@
...
+	}
+
+	var testNodeID NodeID
+	copy(testNodeID[:], "w")
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		err := db.AddItem(testVal[i%len(testVal)], testNodeID)
+
+		if err != nil {
```
Comment (optmzr) [inline store/store_test.go]: Remove this error check

File: store/store_test.go [hunk 16]
```diff
@@ -0,0 +1,263 @@
...
+func TestStoredKeysAdd(t *testing.T) {
+	db, _ := NewDatabase(time.Second*86400, time.Second*3600, time.Second*86400)
+
+	trueHash := [32]byte{174, 79, 167, 92, 82, 249, 190, 142, 129, 67, 178, 149, 52, 212, 158, 150, 67, 136, 83, 10, 170, 233, 83, 34, 158, 194, 62, 241, 14, 168, 19, 103}
+	testVal := "q"
+
+	db.AddLocalItem(trueHash, testVal)
+
+	storedTestKey, err := db.GetLocalItem(trueHash)
+
```
Comment (optmzr) [inline store/store_test.go]: Remove newline

File: store/store_test.go [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+	trueHash := [32]byte{174, 79, 167, 92, 82, 249, 190, 142, 129, 67, 178, 149, 52, 212, 158, 150, 67, 136, 83, 10, 170, 233, 83, 34, 158, 194, 62, 241, 14, 168, 19, 103}
+
+	testVal := "q"
+
+	var testNodeID NodeID
+	copy(testNodeID[:], "w")
+
+	err := db.AddItem(testVal, testNodeID)
+	if err != nil {
```
Comment (optmzr) [inline store/store_test.go]: Remove error check

File: store/store_test.go [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+	fakeHash := [32]byte{17, 69, 167, 92, 82, 249, 190, 142, 129, 67, 178, 149, 52, 212, 158, 150, 67, 136, 83, 10, 170, 233, 83, 34, 158, 194, 62, 241, 14, 168, 19, 103}
+	trueHash := [32]byte{174, 79, 167, 92, 82, 249, 190, 142, 129, 67, 178, 149, 52, 212, 158, 150, 67, 136, 83, 10, 170, 233, 83, 34, 158, 194, 62, 241, 14, 168, 19, 103}
+
+	testVal := "q"
+
+	var testNodeID NodeID
+	copy(testNodeID[:], "w")
+
+	err := db.AddItem(testVal, testNodeID)
+	if err != nil {
```
Comment (optmzr) [inline store/store_test.go]: Remove error check

File: store/store_test.go [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+func TestItemHandler(t *testing.T) {
+	db, _ := NewDatabase(time.Second*0, time.Second*3600, time.Second*86400)
+
+	testVal := "q"
+
+	var testNodeID NodeID
+	copy(testNodeID[:], "w")
+
+	err := db.AddItem(testVal, testNodeID)
+	if err != nil {
```
Comment (optmzr) [inline store/store_test.go]: Remove error check

File: store/store_test.go [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+
+	// After 2 seconds, check if the item  has been removed.
+	timer := time.NewTimer(time.Second * 3)
+	<-timer.C
+
+	_, err = db.GetItem(trueHash)
+	if err == nil {
+		t.Errorf("Item is still in DB after 2 seconds, handler not working.")
+	}
+
```
Comment (optmzr) [inline store/store_test.go]: Remove newline

File: store/store_test.go [hunk 21]
```diff
@@ -0,0 +1,263 @@
...
+}
+
+func TestItemHandlerRepub(t *testing.T) {
+
```
Comment (optmzr) [inline store/store_test.go]: Remove newline

File: store/store_test.go [hunk 22]
```diff
@@ -0,0 +1,263 @@
...
+	db, _ := NewDatabase(time.Second*86400, time.Second*3600, time.Second*0)
+	testVal := "q"
+
+	var testNodeID NodeID
+	copy(testNodeID[:], "w")
+	trueHash := [32]byte{174, 79, 167, 92, 82, 249, 190, 142, 129, 67, 178, 149, 52, 212, 158, 150, 67, 136, 83, 10, 170, 233, 83, 34, 158, 194, 62, 241, 14, 168, 19, 103}
+
+	err := db.AddItem(testVal, testNodeID)
+	if err != nil {
```
Comment (optmzr) [inline store/store_test.go]: Remove error check

File: store/store_test.go [hunk 23]
```diff
@@ -0,0 +1,263 @@
...
+		t.Errorf("some error in AddItem that is not yet defined")
+	}
+
+	trueHash := [32]byte{174, 79, 167, 92, 82, 249, 190, 142, 129, 67, 178, 149, 52, 212, 158, 150, 67, 136, 83, 10, 170, 233, 83, 34, 158, 194, 62, 241, 14, 168, 19, 103}
+
+	storedTestItem, err := db.GetItem(trueHash)
+
```
Comment (optmzr) [inline store/store_test.go]: Remove newline

File: store/store.go (hunk 3 above)
Comment (Hjortsberg) [reply store/store.go]: Makes sense. Done.

File: store/store.go (hunk 2 above)
Comment (Hjortsberg) [reply store/store.go]: Sweet!

File: store/store.go (hunk 5 above)
Comment (Hjortsberg) [reply store/store.go]: Will use this directly =)

File: store/store.go (hunk 7 above)
Comment (Hjortsberg) [reply store/store.go]: Yeah, old artifact.

File: store/store.go (hunk 9 above)
Comment (Hjortsberg) [reply store/store.go]: Theís commenting style is used to be able to autogenerate godoc right?

File: store/store.go (hunk 11 above)
Comment (Hjortsberg) [reply store/store.go]: It was a suspect of the data race before. I had changed this many times and lost track of what i thought should be there. :eyes:

File: store/store.go (hunk 9 above)
Comment (optmzr) [reply store/store.go]: I don't think so... It's still able to generate docs. Just a convention that has been decided upon, some tools will give errors on bad comments

**Pull request merged**